python/cudf/cudf/core/frame.py

```diff
@@ -1213,7 +1208,9 @@ def _validate_merge_cfg(
                 )
 
         # Require same total number of columns to join on in both operands
-        if not (len_left_on + left_index) == (len_right_on + right_index):
+        if not (len_left_on + left_index * len(lhs.index.names)) == (
+            len_right_on + right_index * len(rhs.index.names)
+        ):
```

Comment:
Shouldn't we only need to compare the number of index columns if we're joining against the indexes?

Reply:
There's also the case where we want to join on index columns as well as specifying `on`, or where we want to join the left index columns against right data columns.

Reply:
Either way, it looks like this checks happens in every case instead of conditionally?

Reply:
I see what you're saying. There is a simplicity tradeoff I suppose, I think this one expression gives the right answer in all four cases, the drawback being calling `len(df.index.names)` for the left and right sometimes just to later zero that term out of the equation. Additionally, while three out of the four cases actually need this term, joining on data columns alone is probably the most actually common case. So most of the time, the `len` calls are likely wasted. That said, I think they're pretty cheap.

Reply:
Sorry, is `len(lhs.index.names)` 0 when we're not joining on the indexes? My concern was this was incorrect behavior where say I had a left table with 1 index column and a right table with 3 index columns but was only joining on 1 non-index column.

Reply:
`len(lhs.index.names)` is nonzero, but `len(lhs.index.names) * left_index` is zero when `left_index==False`. So in the case you describe, `len_left_on + left_index * len(lhs.index.names)` would be `0 + True * 1` on the left, and `1 + False * 3` on the right, we'd get `1 == 1` and pass it as a valid merge configuration.

Reply:
Brainfart in `left_index` / `right_index` being booleans. Makes complete sense. Thanks for the explanation!